presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationThroughExchange.java

```diff
@@ -0,0 +1,312 @@
+                Map<Symbol, Symbol> exchangeMap = buildExchangeMap(node.getOutputSymbols(), node.getInputs().get(i));
+                AggregationWithLayout childPartial = generateNewPartial(partial, currentSource, exchangeMap, reuseSymbol);
+                inputs.add(childPartial.getLayout());
+                PlanNode child;
+                if (allowPushThroughChildren) {
+                    child = context.rewrite(currentSource, childPartial.getAggregationNode());
+                }
+                else {
+                    child = context.defaultRewrite(childPartial.getAggregationNode());
+                }
+                newChildren.add(child);
+            }
+            PartitioningScheme partitioningScheme = new PartitioningScheme(
+                    node.getPartitioningScheme().getPartitioning(),
+                    partial.getOutputSymbols(),
+                    partial.getHashSymbol());
+            return new ExchangeNode(
+                    node.getId(),
+                    node.getType(),
+                    node.getScope(),
+                    partitioningScheme,
+                    newChildren,
+                    inputs);
+        }
+
+        private boolean allowPushThrough(PlanNode node)
+        {
+            if (node instanceof ExchangeNode) {
+                ExchangeNode exchangeNode = (ExchangeNode) node;
+                return exchangeNode.getType() != REPLICATE && !exchangeNode.getPartitioningScheme().isReplicateNulls();
+            }
+            return false;
+        }
+
+        private Symbol generateIntermediateSymbol(Signature signature)
+        {
+            InternalAggregationFunction function = functionRegistry.getAggregateFunctionImplementation(signature);
+            return allocator.newSymbol(signature.getName(), function.getIntermediateType());
+        }
+
+        private Map<Symbol, Symbol> buildExchangeMap(List<Symbol> exchangeOutput, List<Symbol> sourceOutput)
+        {
+            checkState(exchangeOutput.size() == sourceOutput.size(), "exchange output length doesn't match source output length");
+            ImmutableMap.Builder<Symbol, Symbol> builder = ImmutableMap.builder();
+            for (int i = 0; i < exchangeOutput.size(); i++) {
+                builder.put(exchangeOutput.get(i), sourceOutput.get(i));
+            }
+            return builder.build();
+        }
+
+        private List<Expression> replaceArguments(List<Expression> arguments, Map<Symbol, Symbol> exchangeMap)
+        {
+            Map<SymbolReference, Symbol> symbolReferenceSymbolMap = new HashMap<>();
+            for (Map.Entry<Symbol, Symbol> entry : exchangeMap.entrySet()) {
+                symbolReferenceSymbolMap.put(entry.getKey().toSymbolReference(), entry.getValue());
+            }
+            return arguments.stream()
+                    .map(expression -> {
+                        if (expression instanceof SymbolReference && symbolReferenceSymbolMap.containsKey(expression)) {
```

Comment:
i dont think we need this predicate?
```java
expression instanceof SymbolReference
```
I think the containsKey should do that implicitly?